### PR TITLE
Add warn_[cpus|gpus]_not_multiple_of and logic

### PR DESCRIPTION
**Pull request by janbridley**
## Description

Add additional keys to warn users when their job is of an unexpected size. Update Frontier default configuration to use the new keys.

## Motivation and context

On clusters with no shared partition, some jobs need to request more resources than strictly required in order for the allocation to be granted. Adding a warning configuration option will give users feedback on underutilized jobs without preventing these partial jobs from running.

Note that warnings are returned *after* the job submission is confirmed.

## How has this been tested?

Existing tests have been updated to use the new keys. Due to the nature of the warnings, it is nontrivial to add new tests. This can be addressed in a later PR.

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/row/blob/trunk/doc/src/developers/contributing.md).
- [x] I agree with the terms of the [**Row Contributor Agreement**](https://github.com/glotzerlab/row/blob/trunk/ContributorAgreement.md).
- [x] My name is on the list of contributors (`doc/src/contributors.md`) in the pull request source branch.
- [x] I have added a change log entry to `doc/src/release-notes.md`.

### CODE REVIEW

**Comment by janbridley** [comment]
If it would be beneficial to add new tests, please let me know! I cannot seem to build Row (even trunk) on Frontier so the current code is untested on that system.

File: doc/src/release-notes.md [hunk 1]
```diff
@@ -1,5 +1,15 @@
 # Release notes
 
+## 0.1.4 (2024-06-13)
+
+*Added:*
+
+* `warn_[cpus|gpus]_multiple_of` key for *clusters.toml*
+
+*Changed*
```
Comment (janbridley) [inline doc/src/release-notes.md]:
```suggestion
*Changed:*
```

File: doc/src/release-notes.md [hunk 2]
```diff
@@ -1,5 +1,15 @@
 # Release notes
 
+## 0.1.4 (2024-06-13)
+
+*Added:*
+
+* `warn_[cpus|gpus]_multiple_of` key for *clusters.toml*
```
Comment (joaander) [inline doc/src/release-notes.md]: These keys are misnamed. The action is not to warn when it *is* a multiple of, but when it *is not*.

```suggestion
* `recommend_[cpus|gpus]_multiple_of` key for *clusters.toml*
```
The word `recommend` matches the word in the warning. `prefer` would also work.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! There are some things to do before merging:

- [x] Name the keys appropriately (see the comment below).
- [x] Document the new keys in `cluster.md`.
- [x] Use the new keys in the `andes` cluster.

I've looked into moving up where the warning is issued. This is difficult because we cannot warn until we know the partition. If we rewrote the code to call `find_partition` before `submit`, then we would either end up calling `find_partition` twice (and seeing the warning twice), or making some big refactors to the code that require a persistent scheduler object. For example, we could implement a `prepare` method that must be called on all actions/directories before `submit`. This isn't a very clean API.

Either case would cause a large change in behavior. Currently `submit` will submit the first `N-1` jobs successfully when the N'th cannot find a partition. If we check all the partitions first, then `submit` would submit 0 jobs in this case.

How much of a problem is it that the warning comes after the confirmation?

File: doc/src/release-notes.md (hunk 2 above)
Comment (janbridley) [reply doc/src/release-notes.md]: Would `warn_*_not_multiple_of be clearer` than either of the above? This is a nitpick at this point, but we may as well

**Comment by janbridley** [comment]
> I've looked into moving up where the warning is issued. This is difficult because we cannot warn until we know the partition. If we rewrote the code to call `find_partition` before `submit`, then we would either end up calling `find_partition` twice (and seeing the warning twice), or making some big refactors to the code that require a persistent scheduler object. For example, we could implement a `prepare` method that must be called on all actions/directories before `submit`. This isn't a very clean API.
> 
> Either case would cause a large change in behavior. Currently `submit` will submit the first `N-1` jobs successfully when the N'th cannot find a partition. If we check all the partitions first, then `submit` would submit 0 jobs in this case.
> 
> How much of a problem is it that the warning comes after the confirmation?

I think it is probably OK - while less than optimal, I think the downsides of addressing the issue outweigh the minor upside. Most clusters have a shared partition, and users on Frontier/Andes (those who would see this warning) are likely to have significant experience anyway.

File: doc/src/release-notes.md (hunk 2 above)
Comment (joaander) [reply doc/src/release-notes.md]: Good point. `warn_*_not` is more explicit about what it does. "Prefer" is vague and seems to indicate that **row** might take some sort of additional action based on this preference.

File: doc/src/release-notes.md [hunk 3]
```diff
@@ -1,6 +1,15 @@
 # Release notes
 
-## 0.1.4 (not yet released)
+## 0.1.4 (2024-06-13)
```
Comment (joaander) [inline doc/src/release-notes.md]:
```suggestion
## 0.2.0 (not yet released)
```

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**